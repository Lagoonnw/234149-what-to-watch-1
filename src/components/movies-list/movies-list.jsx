import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {MovieCard} from '../movie-card/movie-card.jsx';
import {GenreMap, VIDEO_PLAY_DELAY_TIME} from '../../constants/constants';
import {connect} from 'react-redux';
import {ActionCreator} from '../../action';
import {filterMovies} from '../../helpers/filterMovies';

export class MoviesList extends PureComponent {
  constructor(props) {
    super(props);
    props.setMovies();
  }

  render() {
    const {movies} = this.props;

    return (
      <div className="catalog__movies-list">
        {movies.map((movie) => this._getMovie(movie))}
      </div>
    );
  }

  _getMovie(movie) {
    const {onClick, activeItem, onMouseEnter, onMouseLeave} = this.props;
    const isPlaying = movie.id === activeItem;
    const _onMouseEnter = () => onMouseEnter(movie.id, VIDEO_PLAY_DELAY_TIME);

    return (
      <MovieCard
        key={movie.id}
        isPlaying={isPlaying}
        onMouseLeave={onMouseLeave}
        onMouseEnter={_onMouseEnter}
        onClick={onClick}
        {...movie}
      />
    );
  }
}

MoviesList.propTypes = {
  movies: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    poster: PropTypes.string.isRequired
  })),
  setMovies: PropTypes.func.isRequired,
  onClick: PropTypes.func.isRequired,
  onMouseLeave: PropTypes.func.isRequired,
  onMouseEnter: PropTypes.func.isRequired,
  activeItem: PropTypes.oneOfType([PropTypes.number, PropTypes.object])
};

MoviesList.defaultProps = {
  activeItem: null
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  movies: filterMovies(state.movies, GenreMap[state.activeGenre])
});

const mapDispatchToProps = (dispatch) => ({
  setMovies: () => dispatch(ActionCreator.setMovies())
});

export default connect(mapStateToProps, mapDispatchToProps)(MoviesList);

