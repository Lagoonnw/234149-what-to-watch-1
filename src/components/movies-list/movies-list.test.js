import React from 'react';
import renderer from 'react-test-renderer';
import {MoviesList} from "./movies-list";

function createNodeMock(element) {
  if (element.type === `video`) {
    return {};
  }
  return null;
}

const mock = [
  {
    id: 54545454,
    name: `Snatch`,
    imgSrc: `pic.jpg`,
    src: `video.mp4`,
    onMouseLeave: jest.fn(),
    onMouseEnter: jest.fn()
  },
  {
    id: 9898988,
    name: `Pulp Fiction`,
    imgSrc: `pic.jpg`,
    src: `video.mp4`,
    onMouseLeave: jest.fn(),
    onMouseEnter: jest.fn()
  },
  {
    id: 545468787,
    name: `The Witcher`,
    imgSrc: `pic.jpg`,
    src: `video.mp4`,
    onMouseLeave: jest.fn(),
    onMouseEnter: jest.fn()
  }
];

describe(`MoviesList render`, () => {
  test(`Should render MoviesList correctly`, () => {
    const options = {createNodeMock};
    const tree = renderer.create(<MoviesList movies={mock}/>, options).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
