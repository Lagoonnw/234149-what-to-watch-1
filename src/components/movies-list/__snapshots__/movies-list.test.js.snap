// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesList render Should render MoviesList correctly 1`] = `
<div
  className="catalog__movies-list"
>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[MockFunction]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        autoPlay={false}
        controls={false}
        height="175px"
        muted={true}
        poster="pic.jpg"
        src="video.mp4"
        width="280px"
      />
    </div>
    <h3
      className="small-movie-card__title"
      onClick={[Function]}
    >
      <a
        className="small-movie-card__link"
      >
        Snatch
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[MockFunction]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        autoPlay={false}
        controls={false}
        height="175px"
        muted={true}
        poster="pic.jpg"
        src="video.mp4"
        width="280px"
      />
    </div>
    <h3
      className="small-movie-card__title"
      onClick={[Function]}
    >
      <a
        className="small-movie-card__link"
      >
        Pulp Fiction
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseEnter={[Function]}
    onMouseLeave={[MockFunction]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        autoPlay={false}
        controls={false}
        height="175px"
        muted={true}
        poster="pic.jpg"
        src="video.mp4"
        width="280px"
      />
    </div>
    <h3
      className="small-movie-card__title"
      onClick={[Function]}
    >
      <a
        className="small-movie-card__link"
      >
        The Witcher
      </a>
    </h3>
  </article>
</div>
`;
